cmd/virt-handler/virt-handler.go

```diff
@@ -275,18 +289,36 @@ func (app *virtHandlerApp) Run() {
 	stop := make(chan struct{})
 	defer close(stop)
 	factory.Start(stop)
-	cache.WaitForCacheSync(stop, factory.ConfigMap().HasSynced)
+	cache.WaitForCacheSync(stop, factory.ConfigMap().HasSynced, vmiInformer.HasSynced)
 
 	go vmController.Run(10, stop)
 
-	logger.V(1).Infof("metrics: max concurrent requests=%d", app.MaxRequestsInFlight)
+	errCh := make(chan error)
+	go app.runPrometheusServer(errCh, certStore)
+	go app.runConsoleServer(errCh, clusterConfig.GetConsoleServerPort(), consoleHandler)
+
+	// wait for one of the servers to exit
+	<-errCh
+}
+
+func (app *virtHandlerApp) runPrometheusServer(errCh chan error, certStore certificate.FileStore) {
+	log.Log.V(1).Infof("metrics: max concurrent requests=%d", app.MaxRequestsInFlight)
 	http.Handle("/metrics", promvm.Handler(app.MaxRequestsInFlight))
+	errCh <- http.ListenAndServeTLS(app.ServiceListen.Address(), certStore.CurrentPath(), certStore.CurrentPath(), nil)
+}
 
-	err = http.ListenAndServeTLS(app.ServiceListen.Address(), certStore.CurrentPath(), certStore.CurrentPath(), nil)
-	if err != nil {
-		log.Log.Reason(err).Error("Serving prometheus failed.")
-		panic(err)
+func (app *virtHandlerApp) runConsoleServer(errCh chan error, port int, consoleHandler *rest.ConsoleHandler) {
+	ws := new(restful.WebService)
+	ws.Route(ws.GET("/namespaces/{namespace}/vmi/{name}/console").To(consoleHandler.SerialHandler))
+	ws.Route(ws.GET("/namespaces/{namespace}/vmi/{name}/vnc").To(consoleHandler.VNCHandler))
+	restful.DefaultContainer.Add(ws)
+	server := &http.Server{
+		Addr:    fmt.Sprintf("%s:%d", app.ServiceListen.BindAddress, port),
+		Handler: restful.DefaultContainer,
+		// we use migration TLS also for console connections (initiated by virt-api)
+		TLSConfig: app.migrationTLSConfig,
 	}
+	errCh <- server.ListenAndServeTLS("", "")
```

Comment:
Can you tie this server into the server which we already use to serve prometheus metrics? There is no need for another webserver.

Reply:
does the prometheus one use the migration tls configuration as well?

Reply:
Right now it uses afaik an untrusted tls certificate. It can as well use an a little less untrusted one.

Reply:
so how can we configure a server with two different certificates?

Reply:
Not sure I can follow. Just use the migration certificate for everything and don't generate the other insecure one.

Reply:
ah, so changing the prometheus server to use the migration certificate, right?

Reply:
:+1: